CORE-5741 CPX signing tool

kyriathar opened the pull request:
Adds CPx sign command to corda cli package module.

- removes existing CPx signatures before signing.
- has an option to allow multi-signing.
- `isSigningRelated` function for jar entries moved in `packaging:packaging` module to be referenced by both `packaging:packaging-verify` and `tools:plugins:package` modules. 
- extracts out common signing options.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/SignCpb.kt:
```diff
@@ -0,0 +1,71 @@
+    }
+}
+
+/**
+ * Check file exists and returns a Path object pointing to the file, throws error if file does not exist
+ */
+private fun requireFileExists(fileName: String): Path {
+    val path = Path.of(fileName)
+    require(Files.isReadable(path)) { "\"$fileName\" does not exist or is not readable" }
+    return path
+}
+
+/**
+ * Check that file does not exist and returns a Path object pointing to the filename, throws error if file exists
+ */
+private fun requireFileDoesNotExist(fileName: String): Path {
+    val path = Path.of(fileName)
+    require(Files.notExists(path)) { "\"$fileName\" already exists" }
+    return path
+}
```
Are these a duplicate of the functions used by other commands?

If so, can we open up the scope of these functions and use the same functions across commands?

kyriathar replied:
Good point, will extract them out to be re-used.

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/SignCpbTest.kt:
```diff
@@ -0,0 +1,61 @@
+        val createdCpb = Path.of("$tempDir/${CreateCpbTest.CREATED_CPB_NAME}")
+
+        val (createdManifestMainAttributes, createdManifestEntries) = TestUtils.getManifestMainAttributesAndEntries(createdCpb)
+//        val createdCpbSigningFiles = TestUtils.getSigningRelatedJarEntries(createdCpb)
+
+        val signedCpb = Path.of("$tempDir/$SIGNED_CPB_NAME")
+        app = SignCpb()
+        CommandLine(app).execute(
+            createdCpb.toString(),
+            "--file=$signedCpb",
+            "--keystore=${testKeyStore}",
+            "--storepass=keystore password",
+            "--key=signing key 2"
+        )
+
+        val (signedManifestMainAttributes, signedManifestEntries) = TestUtils.getManifestMainAttributesAndEntries(signedCpb)
+//        val signedCpbSigningFiles = TestUtils.getSigningRelatedJarEntries(signedCpb)
+
+        assertEquals(createdManifestMainAttributes, signedManifestMainAttributes)
+        assertTrue(
+            createdManifestEntries.size == signedManifestEntries.size &&
+                    !createdManifestEntries.equals(signedManifestEntries)
+        )
+//        assertTrue(
+//            createdCpbSigningFiles.map { it.name } == (signedCpbSigningFiles.map { it.name }) &&
+//                    createdCpbSigningFiles.all {
+//                        signedCpbSigningFiles.contains(it) &&
+//                            it
+//                    }
+//        )
```
Got some commented out code here

kyriathar replied:
I will add this back in, need to add a check for content for: signature files and signature block files.

kyriathar replied:
This is now added back in, checking also jar entries contents.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/CpxSigner.kt:
```diff
@@ -72,4 +94,55 @@ internal object CpxSigner {
             .generateCertPath(certificateChain)
+}
+
+// TODO The below functions are copied and pasted from `JarVerifier`. We should move them to some signing utils module to be re-used.
```
I think it's worth introducing the dependency on packaging-verify now so we can access the versions of these from JarVerifier.

kyriathar replied:
Done

kyriathar commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt:
```diff
@@ -34,64 +37,43 @@ class CreateCpb : Runnable {
-    @Option(names = ["--storepass", "--password", "-p"], required = true, description = ["Keystore password"])
-    lateinit var keyStorePass: String
+    companion object {
+        private const val MANIFEST_VERSION = "1.0"
```
@barrylapthorn added this const for https://github.com/corda/corda-cli-plugin-host/pull/56#discussion_r923306145.

josephzunigadaly commented on libs/packaging/packaging/src/main/kotlin/net/corda/libs/packaging/Utils.kt:
```diff
@@ -75,3 +75,54 @@ internal fun jarSignatureVerificationEnabledByDefault() = System.getProperty(DEF
 
+
+/** Check whether JAR entry is signing related */
+fun isSigningRelated(jarEntry: JarEntry): Boolean {
```
Please don't move this code into packaging. We need a dependency on packaging-verify for upcoming work. There is no need to move code to avoid that dependency when we know we need it.

kyriathar replied:
Sure, will move it back.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt:
```diff
@@ -104,7 +86,7 @@ class CreateCpb : Runnable {
         val manifest = Manifest()
         val manifestMainAttributes = manifest.mainAttributes
-        manifestMainAttributes[Attributes.Name.MANIFEST_VERSION] = "1.0"
+        manifestMainAttributes[Attributes.Name.MANIFEST_VERSION] = MANIFEST_VERSION
```
Will the Java classes write this if we don't?

kyriathar replied:
I tried commenting it out and it does not seem to be getting added then.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt:
```diff
@@ -34,64 +37,43 @@ class CreateCpb : Runnable {
 
     override fun run() {
-        val unsignedCpb = Files.createTempFile("buildCPB", ".$CPB_FILE_EXTENSION")
+        val unsignedCpb = Files.createTempFile("buildCPB", CPB_FILE_EXTENSION)
```
I'd pass null as the file extension. The temporary file is not a valid CPB and we don't want people finding a `.cpb` file in the temp folder and asking us why its not working.

kyriathar replied:
Good point, done.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/SignCpb.kt:
```diff
@@ -0,0 +1,70 @@
+    @CommandLine.Option(names = ["--sig-file"], description = ["Signature file name"])
+    var sigFile: String = CPB_SIGNER_NAME
+
+    @CommandLine.Option(names = ["--erase-signatures", "-e"], arity = "1", description = ["Erase existing signatures"])
+    var eraseExistingSignatures: Boolean = false
```
The default should be true. We always want to erase existing signatures. We only switch it off for multiple signers and even then it's the second signer onwards.

The naming and description don't reveal what the option is for. It exists to support multiple signers. Maybe the option should be called `--multiple-signatures` or `--add-signer`?

Is it possible to run the command without mentioning this option? It should be possible to run the command without specifying this option.

kyriathar replied:
Done. 

Yes, it is possible to run without this option and in that case, `--multiple-signatures` defaults to true.

josephzunigadaly replied:
If we change the name to invert the meaning, we'd also need to invert the default. So `--multiple-signatures` should default to false. The default needs to erase existing signatures. 

The goal is to sign a file that has been through QA, removing the development or test signatures and applying the release signature.

kyriathar replied:
Woops yes will update.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/SignCpb.kt:
```diff
@@ -0,0 +1,70 @@
+import picocli.CommandLine
+
+@CommandLine.Command(
+    name = "sign-cpb"
```
This would work with any of the three files. We could name this `sign` and use it for `cpk`, `cpb`, and `cpi`

kyriathar replied:
Good point, will change it.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/SignCpb.kt:
```diff
@@ -0,0 +1,70 @@
+        val signedCpbPath = requireFileDoesNotExist(outputSignedCpbFile)
+
+        if (eraseExistingSignatures) {
+            val removedSignaturesCpb = Files.createTempFile("removedSignaturesCpb", CPB_FILE_EXTENSION)
```
I think we should replace the file extension with null. It won't be a valid CPB.

kyriathar replied:
Done

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/SignCpbTest.kt:
```diff
@@ -0,0 +1,185 @@
+}
+
+// After Cpx entry is found in both Cpxs their contents are checked to be equal.
+@Suppress("ComplexMethod", "NestedBlockDepth")
```
I would recommend the extract function refactoring instead of this suppression. It allows us to select some code and pull it out into a new function.

It needs independent parts of the code to work on. The while loop under "check entry contents are equal" is a good candidate. Also the while loops that find the right jar entry appear to be duplicates so I think they could be replaced with a call to one function.

kyriathar replied:
Detekt issues fixed as suggested.

I think it also makes sense to move in `TestUtils` because it is generic enough to be re-used by other tests too if we want to.

josephzunigadaly reviewed (commented):
Looks good. Just a few minor comments.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/SignCpx.kt:
```diff
@@ -0,0 +1,71 @@
+    @CommandLine.Parameters(index = "0", paramLabel = "CPx", description=["path of the input CPx"])
+    lateinit var cpxFile: String
+
+    // TODO check how jarsigner works so that we have the equivalent. If overwriting it at least just log something.
+    @CommandLine.Option(names = ["--sig-file"], description = ["Signature file name"])
+    var sigFile: String = CPx_SIGNER_NAME
```
The jarsigner docs say:

> If no -sigfile option appears on the command line, then the base file name for the .SF and .DSA files is the first 8 characters of the alias name specified on the command line, all converted to uppercase. If the alias name has fewer than 8 characters, then the full alias name is used. If the alias name contains any characters that are not allowed in a signature file name, then each such character is converted to an underscore (_) character in forming the file name. Valid characters include letters, digits, underscores, and hyphens.
>
> https://docs.oracle.com/javase/7/docs/technotes/tools/windows/jarsigner.html

The alias name is `signingOptions.keyAlias` in our code.

kyriathar replied:
Registered [CORE-5921](https://r3-cev.atlassian.net/browse/CORE-5921) to align file naming for signature related files with `jarsigner` when signing across all module.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/SignCpx.kt:
```diff
@@ -0,0 +1,71 @@
+
+    companion object {
+        /** Name of signature within Cpx file */
+        private const val CPx_SIGNER_NAME = "CPx-SIG"
```
I think we should remove this and use the same behaviour as jarsigner, using `signingOptions.keyAlias` as the default.

kyriathar replied:
Same as above (will address in [CORE-5921](https://r3-cev.atlassian.net/browse/CORE-5921)).

josephzunigadaly commented on libs/packaging/packaging-verify/src/main/kotlin/net/corda/libs/packaging/verify/internal/Utils.kt:
```diff
@@ -80,4 +81,55 @@ internal fun <T> List<T>.singleOrThrow(noElementsException: Exception, multipleE
+}
+
+/** Check whether JAR entry is signing related */
+fun isSigningRelated(jarEntry: JarEntry): Boolean {
```
This method is inside an "internal" package but we use it in a different module.

kyriathar replied:
Done

kyriathar-r3 committed:
CORE-5741 CPX signing tool (#1692)

Adds CPx sign command to corda cli package module.

- removes existing CPx signatures before signing.
- has an option to allow multi-signing.
- `isSigningRelated` function for jar entries moved in `packaging:packaging` module to be referenced by both `packaging:packaging-verify` and `tools:plugins:package` modules.
- extracts out common signing options.
